istioctl/pkg/multicluster/remote_secret.go

```diff
@@ -0,0 +1,270 @@
+// Copyright 2019 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicluster
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/apimachinery/pkg/runtime/serializer/versioning"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth" // to avoid 'No Auth Provider found for name "gcp"'
+	"k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/client-go/tools/clientcmd/api/latest"
+
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/secretcontroller"
+)
+
+var (
+	Codec  runtime.Codec
+	Scheme *runtime.Scheme
+)
+
+func init() {
+	Scheme = runtime.NewScheme()
+	utilruntime.Must(v1.AddToScheme(Scheme))
+	yamlSerializer := json.NewYAMLSerializer(json.DefaultMetaFactory, Scheme, Scheme)
+	Codec = versioning.NewDefaultingCodecForScheme(
+		Scheme,
+		yamlSerializer,
+		yamlSerializer,
+		v1.SchemeGroupVersion,
+		runtime.InternalGroupVersioner,
+	)
+}
+
+type Options struct {
+	ServiceAccountName string
+	Contexts           []string
+
+	// inherited from root command
+	Namespace  string
+	Kubeconfig string
+}
+
+var (
+	defaultSecretPrefix       = "istio-pilot-remote-secret-"
+	DefaultServiceAccountName = "istio-pilot-service-account"
+)
+
+// NewCreatePilotRemoteSecretCommand creates a new command for joining two clusters togeather in a multi-cluster mesh.
+func NewCreatePilotRemoteSecretCommand(kubeconfig, namespace *string) *cobra.Command {
+	o := Options{
+		ServiceAccountName: DefaultServiceAccountName,
+	}
+
+	c := &cobra.Command{
+		Use:   "create-pilot-remote-secrets <cluster_context>... [OPTIONS]",
+		Short: "Create secret(s) for Pilot to discovery and join remote k8s service registries",
+		Example: `
```

Comment:
Create a secret to allow Istio to use remote k8s ApiServers ? 

Pilot is short term, and the utility has no control over how the SA is configured and what it can do. Based on role bindings we can use the remote access to the SA for other purposes.